### PR TITLE
Use lunchbox plugins for backends

### PR DESCRIPTION
**Pull request by eile**

### CODE REVIEW

**Comment by eile** [comment]
crossref Eyescale/Lunchbox#294

File: keyv/Map.h [hunk 1]
```diff
@@ -256,7 +245,9 @@ class Map
     Map( const Map& ) = delete;
     Map& operator = ( const Map& ) = delete;
 
-    std::unique_ptr< detail::Plugin > _impl;
+    struct Impl;
+    std::unique_ptr< Impl > _impl;
+    std::unique_ptr< Plugin > _plugin;
```
Comment (tribal-tec) [inline keyv/Map.h]: why plugin not in impl?

Comment (eile) [inline keyv/Map.h]: tunnel vision. Fixing.

Comment (eile) [inline keyv/Map.h]: done.

File: keyv/Ceph.cpp [hunk 2]
```diff
@@ -17,33 +17,33 @@
...
+#include <keyv/Plugin.h>
 
+#include <lunchbox/pluginRegisterer.h>
 #include <rados/librados.hpp>
 
 namespace keyv
 {
-namespace ceph
-{
+class Ceph
```
Comment (hernando) [inline keyv/Ceph.cpp]: Missing ;

File: keyv/Ceph.cpp [hunk 3]
```diff
@@ -17,33 +17,33 @@
...
+
 namespace
 {
+lunchbox::PluginRegisterer< Ceph > registerer;
+
 static void _throw( const std::string& reason, const int error )
```
Comment (hernando) [inline keyv/Ceph.cpp]: Static not needed.

File: keyv/Plugin.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    typedef Plugin InterfaceT;
+
+    /** @internal Needed by the PluginRegisterer. */
+    typedef servus::URI InitDataT;
+
+    /**
+     * Set the maximum number of non-commited writes.
+     *
+     * @param size the maximum write queue size
+     * @return the effectice size used by the implementation
```
Comment (tribal-tec) [inline keyv/Plugin.h]: typo 'effectice'

File: keyv/Plugin.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    virtual ~Plugin() {}
+
+    /** @internal Needed by the PluginRegisterer. */
+    typedef Plugin InterfaceT;
+
+    /** @internal Needed by the PluginRegisterer. */
+    typedef servus::URI InitDataT;
+
+    /**
+     * Set the maximum number of non-commited writes.
```
Comment (tribal-tec) [inline keyv/Plugin.h]: typo 'commited'. Besides, should be 'flushed' as there is a flush() function to explicitly commit all pending inserts (or fix flush() doxygen accordingly).

File: keyv/Plugin.h [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    virtual bool insert( const std::string& key, const void* data,
+                         size_t size ) = 0;
+
+    /** Commit all pending insert operations */
+    virtual bool flush() = 0;
+
+    /** @return a value for a key, or an empty string. */
+    virtual std::string operator [] ( const std::string& key ) const = 0;
+
+    /** Get the given values and call a function on each key. */
```
Comment (tribal-tec) [inline keyv/Plugin.h]: kind of sparse documentation. why not just referencing the one from Map.h with ``` @copydoc ``` ?

**Comment by eile** [comment]
both done

File: keyv/Plugin.h (hunk 5 above)
Comment (eile) [inline keyv/Plugin.h]: done.

File: keyv/Plugin.h (hunk 4 above)
Comment (eile) [inline keyv/Plugin.h]: done.

File: keyv/Plugin.h (hunk 6 above)
Comment (eile) [inline keyv/Plugin.h]: done

**Comment by hernando** [comment]
The changes that tribaltec requested seem to be lost in translation.

File: keyv/Plugin.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    virtual bool insert( const std::string& key, const void* data,
+                         size_t size ) = 0;
+
+    /** Commit all pending insert operations */
+    virtual bool flush() = 0;
+
+    /** @return a value for a key, or an empty string. */
+    virtual std::string operator [] ( const std::string& key ) const = 0;
+
+    /** @copydoc Map::getValues */
```
Comment (tribal-tec) [inline keyv/Plugin.h]: Hmm, copydoc all methods here?

**Comment by eile** [comment]
done.

File: keyv/Plugin.h (hunk 7 above)
Comment (eile) [inline keyv/Plugin.h]: done

**Pull request merged**